### DISCUSSION TITLE
Quests contrib

### DISCUSSION THREAD

**Discussion by Chetral** [evennia/evennia#3559] [Evennia Questions] [answered]
Hello Again.
I copied the contrib.tutorial.evadventure.quest in my type classes. Only modification so far is: removing evadventure string from names.
Then i've done a simple quest in two steps.

In the mobs type classes i added this:
```py 
            if "certo" in message.split(' ') or "si" in message.split(' ') or "accetto" in message.split(' '):
                if self.ndb.wait_accept:
                    if self.ndb.wait_accept == caller:
                        self.execute_cmd(f"di {caller.name}: {caller.name.capitalize()} vi ringrazio! Attenderò il vostro ritorno.")

                        caller.quests.add(LostItem)
                        q = caller.quests.get('LostItem')
                        q.add_data('client', self)
                        del self.ndb.wait_accept
                        return
```
The mob interact with the player, if player answer yes to the last question, it gave him the quest, adding it to the caller.quests (the handler). I also add the "client" as the person/object that gave the quest to the player.

So far so good! I go to the mob, do the interactions, get the quest:
<img width="950" alt="image" src="https://github.com/evennia/evennia/assets/55252937/eb29b88a-ce06-4dc8-aea9-6c0b6bfc96df">
(last line is the answer from CmdQuest - mission name, mission status and mission committer).

But, if i stop the server and restart, after doin CmdQuest:
<img width="950" alt="image" src="https://github.com/evennia/evennia/assets/55252937/720325ea-4b34-4365-ba50-8a4a0a40bd0d">

For what i can see, the quest was correct:
<img width="950" alt="image" src="https://github.com/evennia/evennia/assets/55252937/999e71eb-b242-40d7-92da-8db10a0b54e4">

What could i have done?

**Comment by Griatch** [comment]
Hm, it looks ok to me - you add the quest class `LostItem` to the questhandler and then adds some extra stuff to it :thinking: It fails to get the quest class, so the issue would be in the `LostItem` class somehow. Could you share how that class looks too?

**Reply by Chetral** [reply]
Sure, here the complete class:
```py
from typeclasses.quests import Quest
from server.utils import move_coin

class LostItem(Quest):

    key = "LostItem"
    desc = "L'anello perduto"
    
    help_start = "Ritrova l'anello perso nelle fogne"
    help_end = "Riporta l'anello al suo propietario."
    
    def step_start(self, *args, **kwargs):
        obj = [x for x in self.quester.contents if x.tags.has("LostItem", category="quest")]
        if len(obj) > 0:
            self.quester.msg("Hai recuperato l'anello! Vai a riportarlo.")
            self.current_step = "end"
            self.progress()

    def step_end(self, *args, **kwargs):
        if self.quester.location == self.client.location:
            obj = [x for x in self.client.contents if x.tags.has("LostItem", category="quest")]
            if len(obj) > 0:
                obj[0].delete()
                self.client.execute_cmd(f"di {self.quester.name}: Mi hai riportato il mio prezioso anello!")
                self.client.location.msg_contents(f"$You() da alcune monete a $You(targ).", from_obj=self.client,  exclude=[self.quester], mapping={'targ':self.quester})
                self.quester.msg(f"{self.client.name.capitalize()} ti da 2 monete di elettro.")
                self.client.execute_cmd(f"di {self.quester.name}: Eccoti una ricompensa per i tuoi servigi")                
                move_coin(self.quester, 9, 'in')
                self.complete()
```

MOVE_COIN is a function i use to calculate changes and create the money (coins are acutal obj).
The item for the quest is tagged as LostItem, the self.client is passed to the quest after it was created (the code is in the first messages, i think it works as it is assigned correctly and saved in the quests[missioni] dict, and returned it i query for it before restarting the server)

**Comment by Griatch** [comment]
I don't see you doing anything fishy here... The quest system specifically saves the Quest _class_ rather than the instance (and lets you save data separately) in order to be able to save this data safely (avoiding the 'save a hidden db object in another object' issue that can be easy to end up in if saving a custom object otherwise). I don't see why the class loading itself would fail like this. 

Saving `self.client` should work since it's saved in a separate Attribute (as you can see in the `quests[mission]` dict) and that is not what fails to load. It's the class itself that fails which is confusing. 

I'd need to spin this up to see what is going on. If you want to try to debug yourself in the interim, you could try to strip down the code to a minimum and see if it breaks / starts working at a certain point.

**Reply by Chetral** [reply]
Just for remove any doubt: i removed my type classes.quests module and imported directly the contrib one. There is the same error.

**Comment by Griatch** [comment, marked as answer]
@Chetral Ok. It sounds like this may be something on the Evennia side. To properly track this, can you please [make an issue](https://github.com/evennia/evennia/issues/new/choose) for this where you include the minimum code needed to reproduce?

**Reply by Chetral** [reply]
sure!
